docs/terrain.md

```diff
@@ -0,0 +1,123 @@
+# Roblox Terrain Binary Format
+
+This document describes the Terrain binary format. In this format there is no field for Version, so it is assumed that any future changes will be additions to the existing format or a new format entirely. This specification does not include the adjacent PhysicsGrid binary format, only SmoothGrid.
+
+# Contents
+
+- [Document Conventions](#document-conventions)
+- [File Structure](#file-structure)
+- [Data Types](#data-types)
+  - [Chunk](#chunk)
+  - [Voxel](#voxel)
+  - [Voxel.Flag](#voxel.flag)
+  - [Material](#material)
+
+## Document Conventions
+
+This document assumes a basic understanding of Rust's convention for numeric types. For example:
+
+- `u32` is an unsigned 32-bit integer
+- `f32` is a 32-bit floating point number
+
+All numeric types are little endian. Floats are stored as a dividend of their type's numeric maximum; as such, a `f8` with a value of `0x31` translates to `0x31 / 0xFF`, or `~0.192`. Floats are unsigned. Signed integers are stored with their absolute value, with preceding bytes containing the signedness information; meaning the range for an `i8` is `-0xFF` -> `0xFF`, instead of `-0x7F` to `0x7F`.
+
+Unless otherwise noted, all structs in this document are assumed to be stored with their components in the sequence listed without any modification. That is, if a struct is listed as being composed of an `i32` and an `f32`, it can be assumed that it is stored as an `i32` followed by an `f32`.
+
+## File Structure
+
+The first two bytes of the blob are `0x01`, which is a magic number, followed by `0x05`, which is a logarithm base 2 of the chunk size in voxels. Currently, the chunk size is 32<sup>3</sup> (32768). Values other than `0x05` are untested.
+
+Immediately following the header is an array of chunks, each of which must contain enough voxels to reach the maximum count (32<sup>3</sup>). Chunks are ascendingly ordered by X, then Y, then Z based on their position in the world. Each chunk represents a cube of 128<sup>3</sup> units in world space.
+
+| Field Name | Format          | Value                                                                 |
+| :--------- | :-------------- | :-------------------------------------------------------------------- |
+| Header     | `[u8; 2]`       | Magic number and chunk size.                                          |
+| Chunks     | [Chunk](#chunk) | As many chunks as needed to fill the blob. Unknown/no maximum amount. |
```

Comment:
I'd prefer the `Header` section be split into two segments:
```suggestion
| Field Name   | Format          | Value                                                                 |
| :----------- | :-------------- | :-------------------------------------------------------------------- |
| Magic Number | `u8`            | The magic number `0x01`                                               |
| Chunk Size   | `u8`            | The size of the chunks logarithm base 2. Should be `0x05`.            |
| Chunks       | [Chunk](#chunk) | As many chunks as needed to fill the blob. Unknown/no maximum amount. |
```

Additionally, how is the `Chunks` field sized? Is it just "until the end of the blob" and you can add as many or as few chunks as you want? This should be written down.